retrieval/discovery/mdns/mdns.go

```diff
@@ -0,0 +1,175 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mdns
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"golang.org/x/net/context"
+
+	"github.com/hashicorp/mdns"
+	"github.com/prometheus/prometheus/config"
+)
+
+const (
+	resolvConf = "/etc/resolv.conf"
+
+	dnsNameLabel = model.MetaLabelPrefix + "mdns_name"
+
+	// Constants for instrumentation.
+	namespace = "prometheus"
+)
+
+var (
+	mdnsSDLookupsCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "mdns_sd_lookups_total",
+			Help:      "The number of mDNS-SD lookups.",
+		})
+	mdnsSDLookupFailuresCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "mdns_sd_lookup_failures_total",
+			Help:      "The number of mDNS-SD lookup failures.",
+		})
+)
+
+func init() {
+	prometheus.MustRegister(mdnsSDLookupFailuresCount)
+	prometheus.MustRegister(mdnsSDLookupsCount)
+}
+
+// Discovery periodically performs DNS-SD requests. It implements
+// the TargetProvider interface.
+type Discovery struct {
+	names []string
+
+	interval time.Duration
+	m        sync.RWMutex
+	port     int
+	qtype    uint16
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *config.MDNSConfig) *Discovery {
+	return &Discovery{
+		interval: time.Duration(10 * time.Second),
+	}
+}
+
+// Run implements the TargetProvider interface.
+func (dd *Discovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	defer close(ch)
+
+	ticker := time.NewTicker(dd.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	dd.refreshAll(ctx, ch)
+
+	for {
+		select {
+		case <-ticker.C:
+			dd.refreshAll(ctx, ch)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (dd *Discovery) refreshAll(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	var wg sync.WaitGroup
+
+	names := []string{
+		"_prometheus-http._tcp",
+		"_prometheus-https._tcp",
+	}
+
+	wg.Add(len(names))
+	for _, name := range names {
+		go func(n string) {
+			if err := dd.refresh(ctx, n, ch); err != nil {
+				log.Errorf("Error refreshing DNS targets: %s", err)
+			}
+			wg.Done()
+		}(name)
+	}
+
+	wg.Wait()
+}
+
+func (dd *Discovery) refresh(ctx context.Context, name string, ch chan<- []*config.TargetGroup) error {
+	fmt.Println("mdns.refresh", ctx, name)
+
+	// Set up output channel and read discovered data
+	responses := make(chan *mdns.ServiceEntry, 100)
+	tg := &config.TargetGroup{}
+	go func() {
+		for response := range responses {
+			labelSet := model.LabelSet{
+				//dnsNameLabel:       model.LabelValue(name),
+				dnsNameLabel:        model.LabelValue(response.Host),
+				model.InstanceLabel: model.LabelValue(strings.TrimRight(response.Host, ".")),
```

Comment:
So I should set something like `model.MetaLabelPrefix + "mdns_hostname"` and friends for this?